firedrake/solving_utils.py

```diff
@@ -106,19 +107,53 @@ def __init__(self, problem, mat_type, pmat_type, appctx=None,
         self.F = problem.F
         self.J = problem.J
 
-        if mat_type != pmat_type or problem.Jp is not None:
+        # For Jp to equal J, bc.Jp must equal bc.J for all EquationBC objects.
+        Jp_eq_J = all(b.Jp_eq_J for b in problem)
+
+        if mat_type != pmat_type or not Jp_eq_J:
             # Need separate pmat if either Jp is different or we want
             # a different pmat type to the mat type.
             if problem.Jp is None:
                 self.Jp = self.J
             else:
                 self.Jp = problem.Jp
         else:
-            # pmat_type == mat_type and Jp is None
+            # pmat_type == mat_type and Jp_eq_J
             self.Jp = None
 
+        if problem.bcs is None:
+            self.bcs_F = None
+            self.bcs_J = None
+            self.bcs_Jp = None
+        else:
+            # For each of (F, J, Jp), we need to make deep
+            # copy if bc objects themselves have .bcs;
+            # see bcs.py.
+            def create_bc_tree(ebc, form_type):
+                ebcsplit = EquationBCSplit(ebc, getattr(ebc, form_type))
+                for bbc in ebc.bcs:
+                    if isinstance(bbc, DirichletBC):
+                        ebcsplit.add(bbc)
+                    elif isinstance(bbc, EquationBC):
+                        ebcsplit.add(create_bc_tree(bbc, form_type))
+                return ebcsplit
```

Comment:
Lots of the "management" of EquationBC objects is scattered around at point of use. This kind of defeats much of the point of having an object for it.

Can we generalise the functionality of this type of code (I note it looks very similar to `get_ebc_tree` above), and make these methods of `EquationBC` (or at least have utility free functions in `bcs.py`).

Reply:
Added a method `reconstruct` to EquationBC/EquationBCSplit.  It removed these tree-visiting recursions.